dicomparser/parsers.bal

```diff
@@ -0,0 +1,368 @@
+        dicom:ByteOrder byteOrder) returns dicom:DataElementValue|dicom:ParsingError {
+    do {
+        // Value parsing logic is based off of Table 6.2-1 in Part 5
+        dicom:DataElementValue value = ();
+
+        // If VR has a fixed value length bytes, and the valueBytes[] length does not match the length 
+        // in the spec, don't try to parse, use valueBytes[] as the value
+        if dicom:FIXED_LENGTH_VALUE_BYTES.hasKey(vr) {
+            if dicom:FIXED_LENGTH_VALUE_BYTES.get(vr) != valueBytes.length() {
+                value = valueBytes;
+                return value;
+            }
+        }
+
+        if vr == dicom:AT {
+            // AT is an ordered pair of 16-bit unsigned integers that is the value of a data element tag
+            value = check parseTag(valueBytes, byteOrder);
+        }
+
+        if dicom:isBytesVr(vr) {
+            value = valueBytes;
+        }
+
+        if dicom:isFloatVr(vr) {
+            value = check dicom:bytesToFloat(valueBytes, byteOrder);
+        }
+
+        if dicom:isIntVr(vr) {
+            value = check dicom:bytesToInt(valueBytes, byteOrder);
+        }
+
+        if dicom:isStringVr(vr) {
+            value = check string:fromBytes(valueBytes);
+        }
+
+        // Remove leading and trailing spaces if insignificant
+        if dicom:SPACE_INSIGNIFICANT_VRs.indexOf(vr) != () {
+            if value is string {
+                value = value.trim();
+            }
+        }
+
+        // If the value is of string type
+        // Remove leading and trailing spaces
+        if value is string {
+            value = value.trim();
+        }
+
+        return value;
+    } on fail error e {
+        return error dicom:ParsingError(string `Value parsing failed: ${valueBytes.toString()}`, e);
+    }
+}
+
+# Parses a DICOM sequence value.
+#
+# + 'source - The source of the sequence value to be parsed
+# + transferSyntax - The transfer syntax of the sequence data
+# + return - A `dicom:SequenceValue` if the parsing is successful, or a `dicom:ParsingError` if the parsing fails
+public isolated function parseSequenceValue(byte[]|io:ReadableByteChannel 'source,
+        dicom:TransferSyntax transferSyntax) returns dicom:SequenceValue|dicom:ParsingError {
+    do {
+        if !isSupportedTransferSyntax(transferSyntax) {
+            fail error(string `Unsupported transfer syntax: ${transferSyntax}`);
+        }
+
+        io:ReadableByteChannel sequenceByteChannel;
+
+        if 'source is byte[] {
+            sequenceByteChannel = check io:createReadableChannel('source);
+        } else { // io:ReadableByteChannel
+            sequenceByteChannel = 'source;
+        }
+
+        // Sequence parsing logic is based off of Section 7.5 in Part 5
+        // Sequence data element value (items) dataset
+        dicom:SequenceValue sequence = table [];
+        dicom:ByteOrder byteOrder = dicom:getByteOrder(transferSyntax);
+
+        while true {
```

Comment:
Can this lead to busy wait? Can't we use any kind of exit condition based on the bytechannel? Seems we are only relying on "SEQUENCE_DELIMITER_TAG"

Reply:
Current implementation only supports parsing SQ data elements encoded with an undefined length, in which case, the only way to know the end of the data element value is to keep parsing until we encounter a sequence delimitation data element. So this logic is needed and should not lead to a busy wait as the loop will break when there's a parsing error.

Added a new check that validates the parsed item tag which acts as another exit condition in https://github.com/ballerina-platform/module-ballerinax-health.dicom/commit/3807383514d097fe372c936aa46ef84a627a58d3.